keps/sig-api-machinery/20190603-immutable-fields.md

```diff
@@ -0,0 +1,539 @@
+  
+  Now a `v2` update is sent with the same object just received. Now the incoming request payload is defaulted via the `v2` schema and hence `foo.a` is set. The `v1` stored object is read and converted to `v2` without setting `foo.a`. The immutability check fails.
+  
+The learning from the example is that the stored version has to have "at least the defaults as any other version". With arbitrary conversions, there is no way though to mechanically check that versions are compatible with immutability. But this is not new as even without immutability, OpenAPI schemas of multi-version CRDs with conversions must be carefully crafted by the developer to work as expected from the API consumer point of view.
+
+### Publishing
+
+The `x-kubernetes-immutable` and `x-kubernetes-immutable-keys` vendor extensions are published via `/openapi/v2` as is.
+
+### Suggested marker syntax
+
+In analogy to `+required`, `+optional` we propose to add a marker to kubebuilder's
+controller-gen named `+immutable`, which covers the whole subtree of the object 
+(compare field-selection section for extension).
+
+```
+// The name can not be changed after creation.
+// +immutable
+Name string
+
+// The list of containers can not change AT ALL after creation.
+// No single field in existing containers can be changed, added or deleted,
+// no new containers can be added, no existing container can be removed.
+// +immutable
+Containers []Containers
+```
+
+The `x-kubernetes-immutable-keys: true` vendor extension is expressed via `// +immutable=keys` in Golang types:
+
+### Future outline sketch: native resources
+
+For native resources we add a pre-immutability-check normalization step for objects 
+decoded from JSON which have normalizations defined:
+
+1. versioned JSON blob comes in a request
+2. unmarshalled into versioned Golang struct
+3. defaulting
+4. conversion to internal
+5. if immutability is enabled for the resource:
+   1. marshalling into JSON in-memory
+   2. normalization creating a copy with shared data-structures
+   3. strict immutability check against the old object, coming from proto assuming it is normalized.
+
+### Future outline: protobuf
+
+Protobuf encoding unifies unset, empty and null values for slices and maps. Those three cases 
+cannot be differentiated an lead to the same on-the-wire value. In native types
+we use a semantic equality to factor those differences out if JSON is used on the wire, but
+protobuf is used for storage. In CRs decoding and encoding is always faithful with the 
+traditional JSON format used on-the-wire and for storage.
+
+When adding protobuf encoding to CRDs in the future, we have to (without major, non-standard
+efforts) identify unset, empty and null for CRs as well. This leads to the idea to use a less
```

Comment:
There is an alternative.  

We could  generate proto files and proto descriptors from the `types.go`.  A `map[string]string` could turn into a `map<StringOrEmptyOrNull, StringOrEmptyOrNull>`.

Reply:
How unusual would such an encoding be (outside of kube-proto) ?

We could derive the right proto schema from the specified normalization behaviour. But then we come to the question of compatibility if you change normalization at a later point.

We should start early to flesh out proto targetting 1.18. There seem to be so many details to get right.

Reply:
Proto's standard libraries define something like this which is used for proto's JSON support:
https://github.com/protocolbuffers/protobuf/blob/master/src/google/protobuf/struct.proto

We'd basically be defining something similar.

Reply:
Ic, just another sum type (oneof). The question will be whether we then model something in Proto that was just an accident by our use of JSON initially. With JSON we have no choice to get null and undefined as separate values, with proto we could fix our accident.

Reply:
> With JSON we have no choice to get null and undefined as separate values, with proto we could fix our accident.

We need to find a good balance between the "API semantics make sense", "APIs have too many knobs" and "consistency between CRD and built-in types" trade-off.

Reply:
Some further thoughts for proto and normalization:

- We already have `nullable` to explicitly allow nulls.
- With native types we have `omitempty` which maps null and empty to undefined.

1. In the case of non-nullable, we only have two values. Without normalization we get:
   ```
   undefined -> undefined
   empty -> empty
   ```
   By using a default, we can get:
   ```
   undefined -> empty
   empty -> empty
   ```
   Omitempty would give us:
   ```
   undefined -> undefined
   empty -> undefined
   ```
   Hence, if we had `x-kubernetes-omitempty: true` we would have covered all variants we know 
   today with native types.

2. With nullable and no normalization we get
   ```
   undefined -> undefined
   null -> null
   empty -> empty
   ```
   With defaults, we are able to express:
   ```
   undefined -> null
   null -> null
   empty -> empty
   ```
   With omitempty, we would get:
   ```
   undefined -> undefined
   null -> undefined
   empty -> undefined
   ```
   What we miss is:
   ```
   undefined -> empty
   null -> empty
   empty -> empty
   ```
   I question that we really need this at all. If we identity empty in proto with undefined in JSON, `x-kubernetes-omitempty: true` would be enough.

In other words, we maybe only need `x-kubernetes-omitempty: true` and cover all interesting cases.